core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/EntityStatisticsCacheMonitor.scala

```diff
@@ -0,0 +1,112 @@
+package org.broadinstitute.dsde.rawls.monitor
+
+import akka.actor._
+import akka.pattern._
+import cats.effect.{ContextShift, IO}
+import cats.implicits._
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.rawls.dataaccess.SlickDataSource
+import org.broadinstitute.dsde.rawls.monitor.EntityStatisticsCacheMonitor.{EntityStatisticsCacheMessage, HandleBacklog, Sweep}
+import slick.dbio.DBIO
+
+import java.sql.Timestamp
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+
+object EntityStatisticsCacheMonitor {
+  def props(datasource: SlickDataSource, limit: Int)(implicit executionContext: ExecutionContext, cs: ContextShift[IO]): Props = {
+    Props(new EntityStatisticsCacheMonitorActor(datasource, limit))
+  }
+
+  sealed trait EntityStatisticsCacheMessage
+  case object Sweep extends EntityStatisticsCacheMessage
+  case object HandleBacklog extends EntityStatisticsCacheMessage
+}
+
+class EntityStatisticsCacheMonitorActor(val dataSource: SlickDataSource, val limit: Int)(implicit val executionContext: ExecutionContext, val cs: ContextShift[IO]) extends Actor with EntityStatisticsCacheMonitor with LazyLogging {
+
+  override def preStart(): Unit = {
+    super.preStart()
+    self ! Sweep
+  }
+
+  override def receive = {
+    case Sweep => sweep() pipeTo self
+    case HandleBacklog => handleBacklog() pipeTo self
+  }
+
+}
+
+trait EntityStatisticsCacheMonitor extends LazyLogging {
+
+  implicit val executionContext: ExecutionContext
+  implicit val cs: ContextShift[IO]
+  val dataSource: SlickDataSource
+  val limit: Int
+
+  def sweep() = {
+    val checkFuture = for {
+      recordsToUpdate <- dataSource.inTransaction { dataAccess =>
+        dataAccess.workspaceQuery.listOutdatedEntityCaches(limit)
+      }
+
+      updateResults <- recordsToUpdate.toList.traverse { case (workspaceId, lastModified) =>
+        IO.fromFuture(IO(updateStatisticsCache(workspaceId, lastModified))).map { _ =>
+          (workspaceId, lastModified.getTime, System.currentTimeMillis())
+        }
+      }.unsafeToFuture()
+    } yield {
+      updateResults.foreach { case (workspaceId, lastModified, cacheUpdated) =>
+        logger.info(s"Updated workspace $workspaceId. Cache was ${cacheUpdated - lastModified} millis out of date")
+      }
+      logger.info(s"Sweep complete. Updated entity cache for ${recordsToUpdate.length} workspace(s), with the limit set to $limit")
+      if(recordsToUpdate.length < limit) HandleBacklog
+      else Sweep
+    }
```

Comment:
I'm worried about a case where we have a problem workspace. I'm not sure what condition would cause this, but if we have a workspace that consistently fails to calculate its cache ... the monitor can get stuck on that workspace, continually trying and failing to process it, rolling back the transaction, and moving on to the next sweep ... in which the same problem workspace is again processed. We'd be stuck in a loop.

Reply:
For a workspace like this - could we log that it failed to process and set its cache values to null or 0?

Reply:
I'm trying to think of a state that would allow us to mark these workspaces as ignored. Null or 0 will still cause them to be picked up by the monitor. I think setting the time to the future could technically work with some modified logic but that feels very cryptic to me so I'm not going to do that. This requires some more thought.

Reply:
Thought about this a bunch more. I ended up going with Anu's suggestion of using 0 as the timestamp. I was a bit hesitant to use a magic/cryptic value but I can't think of any other way to possibly represent an ignored workspace given the way that cache state is currently represented. I really want to avoid complicating the implementation by adding another database column or introducing more comprehensive states. IMO everything should be driven off of timestamps. Hopefully these cases are extremely rare. At least in these cases, the endpoint will just default to the non-cached version and still return the correct result.

LMK what you think.

Reply:
I think this is fine - I would hope that a workspace getting into a problematic state where the cache fails to calculate should be extremely rare. We should document somewhere in the code what a 0 timestamp designates, as well.